### ISSUE TITLE
BUG: 'zh' localization directory not exist

### ISSUE THREAD

**Issue by IsQiao** [OrchardCMS/OrchardCore#11672]
### To Reproduce
1. when my chrome broswer language is '中文(简体)', the `CurrentUICulture` field value is `zh`,
2. then ModularPoFileLocationProvider()  can't find the `zh` localization dir. because it is not existed.
3. so the website always display `en-US` culture.

### Expected behavior
show Chinese properly

### Screenshots
![image](https://user-images.githubusercontent.com/8536046/167640053-6e40598b-9eca-4229-904e-7f80a2cb3301.png)
![image](https://user-images.githubusercontent.com/8536046/167640564-c259616f-b23f-438e-867b-1ef7c3a7bc32.png)
![image](https://user-images.githubusercontent.com/8536046/167641150-234b2f2b-fcea-4aec-ad34-b66f8d49f26d.png)
![image](https://user-images.githubusercontent.com/8536046/167641484-e085f1e5-cc4f-475c-8b9d-a437ba96b648.png)
![image](https://user-images.githubusercontent.com/8536046/167641642-8c24bbc4-adfb-464d-aa43-f304333d7aaa.png)

**Comment by Skrypt** [comment]
There are discrepancies between different OS with these culture codes. You simply need to rename the folder of your choice to the appropriate culture code for this to work.

**Comment by IsQiao** [comment]
rename the directory can resolve it, but There are same problems in macos and windows 10. 
so i suguest to add `zh` dir to `OrchardCore.Translations` repo, or add some codes to enhance it.

**Comment by Skrypt** [comment]
@agriffard Does Crowdin supports adding the "zh" culture or do we keep only specific culture/regions culture codes?

**Comment by hyzx86** [comment]
It looks like it's just a display problem here, because it works fine on setup page, but once you switch to English, you can't change it back to Chinese

The option `zh-CN` is not listed  here

But I remember that there was no problem with some previous versions, I wonder why `zh-CN` was deleted

@agriffard

**Comment by hishamco** [comment]
FYI Crowdin nowdays have both `zh-CN` and `zh-TW`, so when you choose `zh` can't find the root culture, so you can choose one of the supported cultures or rename the culture folder as workaround

**Comment by hyzx86** [comment]
@hishamco No, that is not a right way.
As a Web application, we should follow the browser Settings, not let the user choose

![image](https://user-images.githubusercontent.com/15613121/210248263-4c2ed708-dbc7-4039-8093-ddec3f5833db.png)

**Comment by hishamco** [comment]
FYI localization APIs already have 3/4 `RequestCultureProvider`s, then it uses a fallback mechanism to the parent culture in case the current cultures is not present

**Comment by hyzx86** [comment]
Yes, I think of a problem, language choice, on the Setup page is good, there use 
`IOptions<RequestLocalizationOptions> LocOptions`

let me have a try..

**Comment by hyzx86** [comment]
faild 🤣
![image](https://user-images.githubusercontent.com/15613121/210249740-b9482a1e-2566-4e4c-ad8c-3270efa52634.png)

``` C#
model.Cultures = _locOptions.Value.SupportedCultures 
                        .Select(cultureInfo =>
                        {
                            return new CultureEntry
                            {
                                Supported = settings.SupportedCultures.Contains(cultureInfo.Name, StringComparer.OrdinalIgnoreCase),
                                CultureInfo = cultureInfo,
                                IsDefault = String.Equals(settings.DefaultCulture, cultureInfo.Name, StringComparison.OrdinalIgnoreCase)
                            };
                        }).ToArray();
```

**Comment by hishamco** [comment]
Just let me know what will happen if you set your culture to `zh-CN` in the admin?

**Comment by hyzx86** [comment]
> FYI localization APIs already have 3/4 `RequestCultureProvider`s, then it uses a fallback mechanism to the parent culture in case the current cultures is not present

Yes, my layer tried to get all cultures first using `CultureInfo.GetCultures(CultureTypes.AllCultures)` and then all subcultures through `SelectMany(x=>x.getAllChildxxx)`, but I couldn't find the right way .

**Comment by hyzx86** [comment]
> Just let me know what will happen if you set your culture to zh-CN in the admin?

https://user-images.githubusercontent.com/15613121/210253277-a335ff13-e180-4b76-8c5f-9e6d3dd86e61.mp4

But, there is no place for switching languages in the OC administration topic.

**Comment by hyzx86** [comment]
![chrome_jWII5d0t2D](https://user-images.githubusercontent.com/15613121/210253715-f0adafa4-c6ee-4138-8d95-48be8299ac79.gif)

**Comment by hishamco** [comment]
Try to set `zh-CN` as default culture and have a try

**Comment by hishamco** [comment]
BTW I think you did the following fallback

https://github.com/OrchardCMS/OrchardCore/blob/083d5d7f9885b6ba034f642259cb94326af9e551/src/OrchardCore/OrchardCore.Localization.Core/DefaultPluralRuleProvider.cs#L66-L78

Right?

**Comment by hyzx86** [comment]
> BTW I think you did the following fallback
> 
> https://github.com/OrchardCMS/OrchardCore/blob/083d5d7f9885b6ba034f642259cb94326af9e551/src/OrchardCore/OrchardCore.Localization.Core/DefaultPluralRuleProvider.cs#L66-L78
> 
> Right?

Yes, we can only get parents from children, but not vice versa. I haven't found a way

**Comment by hishamco** [comment]
Of course you can't ;)

**Comment by hyzx86** [comment]
> Try to set zh-CN as default culture and have a try

I don't need to set `zh-CN` as Default, because the program recognizes that the browser's default language is
`zh-CN`, so when I add `zh-CN`, OC will automatically apply it

**Comment by hishamco** [comment]
To make it clear, what's your browser default language? your OC default language?

**Comment by hyzx86** [comment]
> ![image](https://user-images.githubusercontent.com/15613121/210248263-4c2ed708-dbc7-4039-8093-ddec3f5833db.png)

in my submission.. All the billions of browsers in China should be `zh-CN` 🤣

**Comment by hishamco** [comment]
So, what's the problem then? Just set your default culture in OC to `zh-CN` then it should work

**Comment by hyzx86** [comment]
problem is ..
![image](https://user-images.githubusercontent.com/15613121/210264463-13a7ca6e-c153-4475-b5bc-d33358f203a6.png)

**Comment by hyzx86** [comment]
![image](https://user-images.githubusercontent.com/15613121/210264515-422a9694-5a78-434c-bc69-293d3ce0820c.png)

**Comment by hyzx86** [comment]
As I did in PR, zh CN can only be obtained in this way

![image](https://user-images.githubusercontent.com/15613121/210264693-c6038c40-8ae2-452c-9685-7fd55022f5b2.png)

**Comment by hishamco** [comment]
leave the code for now , the cultures might differ from OS to another, just let me know if you set `zh-CN` as default culture in OC what happen? is it displayed correctly or not?

**Comment by hyzx86** [comment]
in .net 5.0
![image](https://user-images.githubusercontent.com/15613121/210264882-f12dec4f-1d9d-49f7-80a3-07056f7b7f27.png)

**Comment by hyzx86** [comment]
> rename the directory can resolve it, but There are same problems in macos and windows 10. so i suguest to add `zh` dir to `OrchardCore.Translations` repo, or add some codes to enhance it. (_^_^_)

As mentioned above, it is the same on Mac and Windows

> leave the code for now , the cultures might differ from OS to another, just let me know if you set zh-CN as default culture in OC what happen? is it displayed correctly or not?

correctly!

**Comment by hishamco** [comment]
If it's displayed correctly, the `zh` is not a problem because OC don't provide a resources for it by default

**Comment by hyzx86** [comment]
Ok, so we should change the zh-cn folder to or map it to zh in oc

then  zh-CN -> zh-Hans -> zh

**Comment by hishamco** [comment]
if we change it to `zh` what about `zh-TW`?

> then zh-CN -> zh-Hans -> zh

For that I told you try to set your default culture to `zh-CN` and forgot everything about zh :)

**Comment by hyzx86** [comment]
But if I do not explicitly add `zh-CN` there, there will be no `zh-CN` option on the page
Therefore, 'zh CN' cannot be set as the default

**Comment by hishamco** [comment]
Did you mean there's no option for `zh-CN` in the available cultures?

**Comment by hyzx86** [comment]
😥My PR just add an option on the UI

**Comment by hishamco** [comment]
@hyzx86 please refer to https://learn.microsoft.com/en-us/dotnet/api/system.globalization.cultureinfo?view=net-7.0#cultureinfo-and-cultural-data

`zh-CN` is an alias to `zh-Hans-CN` which is should be available in your OC installation

**Comment by hishamco** [comment]
I realize that we need to rename the Chinese cultures folders to `zh` or in case to support simplified & traditional versions we should use `zh-Hans-CN` and `zh-Hant-TW` @hyzx86 correct me if I'm wrong

/cc @sebastienros @agriffard

**Comment by hishamco** [comment]
I realize that we need to rename the Chinese cultures folders to `zh` or in case to support simplified & traditional versions we should use `zh-Hans-CN` and `zh-Hant-TW` @hyzx86 correct me if I'm wrong

/cc @sebastienros @agriffard

**Comment by hyzx86** [comment]
Let me try, but I don't think it should recognize the zh-CN passed in by the browser

**Comment by hishamco** [comment]
I think we should ignore both `zh-CN` and `zh-TW`

**Comment by hyzx86** [comment]
![image](https://user-images.githubusercontent.com/15613121/210426427-289ff13d-21ee-4664-a099-05b2a2bf72e7.png)

**Comment by hishamco** [comment]
I think we should ignore both `zh-CN` and `zh-TW`

**Comment by hyzx86** [comment]
Yes, it doesn't recognize the language information passed in by the browser


https://user-images.githubusercontent.com/15613121/210428219-4b50552b-e54f-44a2-b9bb-45cf210f72ce.mp4

**Comment by hyzx86** [comment]
It's been a long time coming, almost a year since 6.0 was released
This question is very unfriendly to Chinese users
New Chinese developers think there is no Chinese language option at all
Sometimes can inexplicably install a Chinese interface out.

Or, just to keep things simple, let's iterate over all the language packages supported by OC and just show the ones that are supported.
There are so many languages in the world and OC only supports some of them. Why list them all?

**Comment by hishamco** [comment]
![zh-Hans-CN](https://user-images.githubusercontent.com/3237266/210430241-36d4bca2-6abd-4fc3-b545-4c077d78dc1f.png)

As I mentioned above `zh-CN` is not listed that's why I'm suggesting not relaying on such cultures. I just renamed `zh-CN` folder to `zh-Hans-CN` and everything works as expected

**Comment by hyzx86** [comment]
That's because your browser's default language is not zh-CN
I think default should refer to the browser language first and the site language second
As you can see in my screenshot above, if you add the zh-CN option, it will recognize the installed language based on the browser language even if zh-CN is not set as the default language

**Comment by hyzx86** [comment]
> > Just let me know what will happen if you set your culture to zh-CN in the admin?
> 
>  chrome_mD27wx9tZM.mp4 
> But, there is no place for switching languages in the OC admin theme.

here

**Comment by hishamco** [comment]
FYI if your browser has one of the listed culture it will get a priority, coz `RequestCultureProvider` check `AcceptLanguageHeaders` first

**Comment by Skrypt** [comment]
@hyzx86 The reason is that cultures can have different codes based on the OS that you are using. We are using Crowdin to generate these culture folders so that we have at least translations for common culture codes. We can't provide every folder for every culture code.

**Comment by sebastienros** [comment]
What I see is that zh-CN is not available after .NET 5.0 and that might be the source of the problem. So adding it manually might "fix" the issue. Can someone look for the recording of the meeting when we decided which zh languages we would support by default (there were concrete arguments).

Also having support for `zh` might be a good safety net. However I am not sure which current files should become the `zh` ones. Traditional Chineese?

**Comment by sebastienros** [comment]
Results in sharplab io whatever the .NET version

```
zh - Chinese
zh-Hans - Chinese (Simplified)
zh-Hans-CN - Chinese (Simplified, China)
zh-Hans-HK - Chinese (Simplified, Hong Kong SAR China)
zh-Hans-MO - Chinese (Simplified, Macao SAR China)
zh-Hans-SG - Chinese (Simplified, Singapore)
zh-Hant - Chinese (Traditional)
zh-Hant-HK - Chinese (Traditional, Hong Kong SAR China)
zh-Hant-MO - Chinese (Traditional, Macao SAR China)
zh-Hant-TW - Chinese (Traditional, Taiwan)
```

@hyzx86 Assuming we create a new `zh` culture so we have a fallback in all cases (which I think would be safer). Should this `zh` culture use `zh-Hant - Chinese (Traditional)` or `zh-Hans - Chinese (Simplified)` ?
Then we would just add the other one as a specific one
Examples:
1- `zh` (with Simplified values) + `zh-Hant`
2- `zh` (with Traditional values) + `zh-Hans`


Assuming we only keep

**Comment by hyzx86** [comment]
zh-Hant，zh-Hans  
Yes, I think these are enough, because in terms of characters, Chinese is not so much variety, it is more about pronunciation

**Comment by hyzx86** [comment]
But t for me personally... zh-Hans This kind of writing is a little strange. I've been using computers for decades and I'm most familiar with zh-CN because that's how almost all websites are written.
Just like here:
![image](https://user-images.githubusercontent.com/15613121/210865836-b32adba8-4a24-4f85-99e6-49f09ced1ac8.png)

or here
![image](https://user-images.githubusercontent.com/15613121/210865969-c9b755be-7f3d-43a0-84df-1335ee2a8977.png)

**Comment by Skrypt** [comment]
So basically "simplified" Chinese is more common?

**Comment by hyzx86** [comment]
> So basically "simplified" Chinese is more common?

I can't say that because I live in mainland China,
Hong Kong, Macau, Taiwan, Singapore and other places should use more traditional characters.

But I also specifically to check the relevant articles. About whether Taiwanese can read simplified Chinese in mainland China
They say the subtitle translation teams in mainland China are so helpful that many American shows can find subtitles in simplified Chinese the next day..🤣

**Comment by hishamco** [comment]
> What I see is that zh-CN is not available after .NET 5.0 and that might be the source of the problem

I notice that too

> zh-Hant，zh-Hans
> Yes, I think these are enough, because in terms of characters, Chinese is not so much variety, it is more about pronunciation

I totally agree with @hyzx86, that's what I mentioned earlier [here](https://github.com/OrchardCMS/OrchardCore/issues/11672#issuecomment-1370126023)

**Comment by sebastienros** [comment]
More context about how the issue appeared: https://github.com/dotnet/msbuild/issues/7331
Recommendation is to use `zh-Hans` and `zh-Hant`, but since the browsers send `zh-CN` it seems counterproductive. And yes, `zh-cn` is an alias for `zh-Hans-CN` so the "simplified" one.

Right now we have folders for 
- zh-CN
- zh-Hans-CN
- zh-TW

So should we 
- Move `zh-CN` and `zh-Hans-CN` to `zh`
- And move `zh-TW` to `zh-Hant`
=> if someone asks for `zh-CN` from browsers they will get `zh` which is the same simplified version.

**Comment by sebastienros** [comment]
A mitigation is to set this ENV: `DOTNET_SYSTEM_GLOBALIZATION_USENLS=1`
Then the `zh-CN` will appear in the list.

**Comment by hishamco** [comment]
> Right now we have folders for

> zh-CN
> zh-Hans-CN
> zh-TW

We already have `zh-CN` and `zh-TW`, my suggestion to rename them in crowdin into `zh-Hant` and `zh-Hans`. While `zh-CN` is an alias for `zh-Hans-CN`, we could add a special for this while fallback to the parent culture(s).

**Comment by hyzx86** [comment]
Can it be the same as the setup page?

![image](https://user-images.githubusercontent.com/15613121/211052017-60f45826-f9a7-4d97-ac41-1dc05c86ae6d.png)

**Comment by sebastienros** [comment]
Important to note that on the Setup page we add explicitly "zh-CN" and "zh-TW" to the supported cultures and use this list in the dropdown. So it works fine because we don't rely on GetCultures().

I believe the best quick fix is to add zh-CN explicitly in the setting page, by copying it from zh-Hans-CN. This way it will work with our existing files. And reference this issue in the code to remember why we came up with this solution.

**Comment by hyzx86** [comment]
Yes, OC allows users to add languages by placing their own language packs, but why not just check the language folder when open setting page?

Showing all languages isn't very friendly either

**Comment by hishamco** [comment]
What Seb is referring to, is quite similar to what I suggest. Let me create a PR for that

> Yes, OC allows users to add languages by placing their own language packs, but why not just check the language folder when open setting page?

That's good point, but users can add their translations even if they aren't support. So I'm suggesting here to display a warning message because setting non supported culture by OC may not affect the UI translation

**Comment by hishamco** [comment]
> I believe the best quick fix is to add zh-CN explicitly in the setting page, by copying it from zh-Hans-CN

Cloning a culture can't change its name. I just add them explicitly, again I suggest to change the Chinese folder names in Crowdin because `zh-CN` and `zh-TW` are cultures in Windows and alias on Unix systems

I might create another PR preparing for that

**Comment by hishamco** [comment]
FYI there's `CultureAndRegionInfoBuilder` that allow you to register custom culture but I think it's Windows only

**Comment by hishamco** [comment]
I pushed a PR #13049 that should map `zh-CN` -> `zh-Hans-CN` and `zh-TW` -> `zh-Hant-TW` the implementation is extensible to add other culture aliases -if there are-

BTW I release that we need to add both `zn-CN` and `zn-TW` into the cultures list for backward compatibility also to become supported cultures if they user set one of them. Behind the sense OC will map these aliases as expected to `zh-Hans-CN` and `zh-Hant-TW`

The remaining part is to react Crowdin to the naming changes

/cc @agriffard

**Comment by sebastienros** [comment]
@hishamco  I think my suggestion of just adding zh-CN to the GetCultures list is not sufficient as we won't be able to load an actual CultureInfo out of it (it won't be found in the system). So I liked your alias PR.

I believe we will have to do changes in crowding to support `zh` and `zh-Hans-TW`.
`zh` such that we always have a fallback for special chinese cultures. And `zh-Hant` as a fallback for all traditional versions (including zh-TW).

@hishamco do you agree with this?

**Comment by hishamco** [comment]
Nowadays we supports both `zh-CN` & `zh-TW`, so supporting `zh` & `zh-Hant-TW` is solving the problem for Chinese culture nothing but `zh-TW` will be fallback to `zh` which is wrong in this case it should be mapped to `zh-Hant-TW` instead, unless the `zh-Hant-TW` is not supported it should fallback to `zh`

IMHO we need to rename `zh-CN` -> `zh-Hans-CN` & `zh-TW` -> `zh-Hant-TW` which solves the problem. In case of other Chinese cultures such as `zh-SG`, `zh-HK` or `zh-MO` they need to add their localization resources to be localized well, because some of Chinese cultures blown my mind because they could use `zh-Hans` and `zh-Hant` that's why `zh` is not a good option

@hyzx86 you feedback is welcome

**Comment by hishamco** [comment]
> Might need to update the nuget packages script. And maybe better (but not required) to change the setup page default cultures to use zh-Hans-CN and zh-Hant-TW.

Totally agree, I will submit a PR for it

**Comment by agriffard** [comment]
I added the custom languages on crowdin:
https://crowdin.com/project/orchard-core/zh-hans-cn
https://crowdin.com/project/orchard-core/zh-hans-tw
but there is no such thing than "renaming" a culture into another one.

**Comment by hishamco** [comment]
I just see that you create a new one cyrilic and latin cultures instead of renaming

**Comment by hishamco** [comment]
@agriffard is there an option to move the previous translations or shall we do it from the `OC.Translations` repo?

**Comment by sebastienros** [comment]
> IMHO we need to rename zh-CN -> zh-Hans-CN & zh-TW -> zh-Hant-TW which solves the problem. In case of other Chinese cultures such as zh-SG, zh-HK or zh-MO they need to add their localization resources to be localized well, because some of Chinese cultures blown my mind because they could use zh-Hans and zh-Hant that's why zh is not a good option

SG,HK,TW are all traditional. If our files were named zh-Hant then they would all work fine without customization or configuration in the mw. And zh-Hant-TW would still work. And they could still have custom files for these.

Same comment for the simplified variations.

**Comment by sebastienros** [comment]
https://sites.psu.edu/symbolcodes/languages/asia/chinese/

```
Language Tags allow browsers and other software to process Chinese text more efficiently. Below are the recommended codes for different scripts

Chinese: zh (the most generic tag, but rarely used)
Mandarin Chinese, Simplified Script: zh-Hans is preferred, but zh-CN may be found on older sites.
Mandarin Chinese, Traditional Script: zh-Hant or zh-Hant-TW (Taiwan) is preferred zh-TW
Pinyin (Mandarin): zh-Latn-pinyin for Mandarin. If the text is not Mandarin,use one [the dialect codes below](https://sites.psu.edu/symbolcodes/languages/asia/chinese/#dialects).
Cantonese (Hong Kong): zh-HK
```

**Comment by hishamco** [comment]
```
zh - Chinese
zh-Hans - Chinese (Simplified)
zh-Hans-CN - Chinese (Simplified, China)
zh-Hans-HK - Chinese (Simplified, Hong Kong SAR China)
zh-Hans-MO - Chinese (Simplified, Macao SAR China)
zh-Hans-SG - Chinese (Simplified, Singapore)
zh-Hant - Chinese (Traditional)
zh-Hant-HK - Chinese (Traditional, Hong Kong SAR China)
zh-Hant-MO - Chinese (Traditional, Macao SAR China)
zh-Hant-TW - Chinese (Traditional, Taiwan)
```

`zh-MO` & `zh-HK` made me crazy because they are using both simplified and traditional ;)

We could use `zh-Hans` & `zh-Hant` by default if we do so, the other cultures will then fallback properly

**Comment by sebastienros** [comment]
I used the Languages Mapping setting in crowdin to change the target file names (I assume this is why we got 3 folders in the repos right now instead of 2) and I can see it producing zh-Hans and zh-Hant files. I assume the PR should be updated soon. It doesn't change the files in crowdin so that's good.

**Comment by sebastienros** [comment]
The Translation repository is now updated:
- Deleted the zh-Hans-CN folder that was obsolete
- Kept zh-CN and zh-TW just such that users who are upgrading won't be broken and have some time to update to zh-Hans and zh-Hant. The subsequent release should get rid of these folders. We need to add this information to the release notes. And maybe have some docs explaining which languages are available to use by default, and to add in the Supported Cultures settings.
- zh-Hans and zh-Hant are now available as NuGet packages (cloudsmith for now)

**Comment by sebastienros** [comment]
@hishamco the package are pushed correctly, but something is wrong and there is an error after/during:
https://github.com/OrchardCMS/OrchardCore.Translations/actions/runs/3906779824/jobs/6675315818

**Comment by hishamco** [comment]
As I can see only `zh-Hant` has been published from Chinese packages which is weird, let me check the script in `OC.Translations` repo

**Comment by jtkech** [comment]
@hyzx86 

Working on a related issue, I took a look at this one, see https://github.com/OrchardCMS/OrchardCore/pull/13131#issuecomment-1399979709.

After some reading it seems that `zh-CN` is an alias name still related to a valid culture but that is no more returned by `GetCultures()`, so the idea is to add it to the supported cultures if `zh-Hans-CN` is already there, this in our existing `AddSupportedCultures()`, the mapping being in the opposite way.

You may want to try the #13131 branch if you still have any issue.

**Comment by hyzx86** [comment]
Yes, last night I tried updating the latest preview version, 17232
Then delete all files in the language folder, and the language pack is not released to this directory after recompiling

And switching the default language doesn't have any effect

**Comment by jtkech** [comment]
Okay, so can you try the specific branch of this PR #13131 

Look at this PR, the branch being `[jtkech/culture-settings](https://github.com/OrchardCMS/OrchardCore/tree/jtkech/culture-settings)`

**Comment by hyzx86** [comment]
@jtkech 
This branch looks fine, even if I delete all the language packages, it will still generate automatically after regenerating.
![image](https://user-images.githubusercontent.com/15613121/214070552-9c3a4adb-1ac8-4d72-816f-505be951ae46.png)

**Comment by hishamco** [comment]
> even if I delete all the language packages, it will still generate automatically after rebuild.

Impossible ;) it should reads the localization resources

@hyzx86 could you please try to print the current culture on the admin dashboard, also let me know your browser culture too, let me check if the culture alias works as expected or not. coz @jtkech said there's no value for it :) I don't think we introduce it for nothing :)

**Comment by hishamco** [comment]
@jetch after I set with @hyzx86 and showed my his solution with the main branch, his browser sends `zh-CN` culture which is not listed in the admin UI, so the `CultureAliasConversionMiddleware` that we introduced recently mapped the current culture to `zh-Hans-CN` which is works as expected. The only issue he faced that the resources aren't localized because they need to be in `zh-Hans-CN` folder instead of the old folder.

FYI @sebastienros added `zh-Hans` and `zh-Hant` on CloudSmith

**Comment by jtkech** [comment]
@hishamco @hyzx86 

I understand the goal of the alias middleware, that's a good idea, but we can't add for example `zh-CN` to the supported cultures, so the previous localization middleware never sets the `CurrentCulture` to `zh-CN`, so the alias middleware never maps it to `zh-Hans-CN`. I tried it by adding `zh-Hans-CN` to the supported cultures, configuring my browser with `zh-CN`, and by putting break points at the right places.

Indeed, if you add `zh-Hans-CN` to the supported cultures and also mark it as the default culture, in that case  `zh-Hans-CN` will be selected as a fallback but not because the alias middleware mapped it, and then yes nothing is translated as we need the missing localization folders.

That's why in my PR in `AddSupportedCultures()`, if `zh-Hans-CN` is there we also add dynamically `zh-CN`, and I could see the aspnet middleware setting the `CurrentCulture` to `zh-CN`. At this point we could use the alias middleware (that here would do something) but it seems that we can load a valid culture by using the `zh-CN` alias name, so for testing I removed the middleware.

---

> let me check if the culture alias works as expected or not. coz @jtkech said there's no value for it :) I don't think we introduce it for nothing :)

As said above, not saying it is not a good idea, but was never doing any mapping.

> his browser sends zh-CN culture

As I did in my tests.

> which is not listed in the admin UI

As I said myself many times.

> so the CultureAliasConversionMiddleware that we introduced recently mapped the current culture to zh-Hans-CN which is works as expected.

Maybe you added `zh-Hans-CN` to the supported cultures and also marked it as the default culture, indeed in that case  `zh-Hans-CN` will be selected as a fallback but not because the alias middleware mapped it, and then yes nothing is translated as we need the missing localization folders.

But based on my tests, if you add another culture marked as Default e.g. `en-us` and keep `zh-Hans-CN` but no more as the Default culture, even if the browser send `zh-CN`, `en-us` is selected and then no mapping will be done.

In that case with my PR `zh-CN` is selected (the culture loaded with this alias name) and translations are done as long as we keep the old related folder.

**Comment by hishamco** [comment]
@jtkech to make it clear, try the following:

1. Add 'ar', 'en', 'zh-Hans-CN' from Orchard Core admin as supported cultures, where `en` is the default one
2. Let the browser sends `zh-CN`
3. The current culture SHOULD be `zh-Hans-CN`

BTW the UI texts might not be chnage because we don't have `zh-Hans-CN`, but if you have it under localization folder the UI should be transalted properly

**Comment by hishamco** [comment]
Tested on `main` branch, also @hyzx86 tested with me yesterday

**Comment by jtkech** [comment]
@hishamco 

> @jtkech to make it clear, try the following:

This is what I already tried but without the `ar` culture, so I added it and re-tried, same result the alias mapping is never done. Maybe in your case `zh-Hans-CN` is already selected by the aspnet middleware that by default recursively check the parent culture (and it may depend on the system), but in that case too no culture mapping would be done.

That said it seems that in both Windows and Unix `zh-CN` doesn't fallback to `zh-Hans-CN`.

> On Windows systems, the parent culture of the "zh-CN" culture is "zh-Hans", and the parent culture of the "zh-TW" culture is "zh-Hant". The parent culture of both these cultures is "zh". On Unix systems, the parents of both cultures are "zh".

Never fully sure, only based on my debug tests, but maybe in your case the browser not only sends `zh-CN` but also `zh-Hans-CN`, which would explain that `AcceptLanguageHeaderRequestCultureProvider` already selects `zh-Hans-CN`, but here too no culture mapping would be done.

Anyway, at least on my Windows machine the alias mapping from `zh-CN` to `zh-Hans-CN` never happens, unless we add `zh-CN` to the `SupportedCultures` as I do dynamically in my testing PR.

Did you tried it by putting a breakpoint in the alias middleware where it mutates the current culture?

If so, another explanation would be that it may depend on the system user settings when getting the `CurrentCulture` if not already set, but normally the aspnet localization middleware always set it.

Anyway, at least on my Windows machine the alias mapping is never done, so it doesn't seem to be a valid fix, but I may be wrong, only based on real debug tests using breakpoints and so on.

**Comment by hishamco** [comment]
> Maybe in your case zh-Hans-CN is already selected by the aspnet middleware that by default recursively check the parent culture (and it may depend on the system), but in that case too no culture mapping would be done.

The `zh-Hans-CN` selected by the culture alias middleware once the current culture is `zh-CN` I might need to share a screenshot as mp4

**Comment by hishamco** [comment]
> Maybe in your case zh-Hans-CN is already selected by the aspnet middleware that by default recursively check the parent culture (and it may depend on the system), but in that case too no culture mapping would be done.

The `zh-Hans-CN` selected by the culture alias middleware once the current culture is `zh-CN` I might need to share a screenshot as mp4

**Comment by hishamco** [comment]
https://user-images.githubusercontent.com/3237266/214438582-a7063ee8-592f-48a6-8b7d-47ea128ed4f9.mp4

FYI the UI in Chrome appears in Chinese from previous test

**Comment by hishamco** [comment]
It will be better if someone else can confirm if this feature is working on `main` branch

**Comment by jtkech** [comment]
Okay interesting, and I assume you also added in the admin settings `zh-Hans-CN` in the supported cultures and not marked as the default culture, as you mentioned, right?

Okay but with the exact same configuration, in my case when putting a break point in the alias middleware the current culture is `en-US`, I can get `zh` if I add it to the supported cultures in the admin settings, but never `zh-CN` that I can't add in the settings, so no mapping is done on my side.

Can you show your admin culture settings? Did you save them, maybe `zh-CN` is still in the database document. Would be good to put a breakpoint in the module startup to see what is added in the supported cultures, just try to understand how it can select a culture which is not in the supported cultures.

> It will be better if someone else can confirm if this feature is working on main branch

Thanks for trusting me ;) But even if it works for someone else, it will not magically start to work on my machine.

**Comment by hishamco** [comment]
> Can you show your admin culture settings? Did you save them, maybe zh-CN is still in the database document.

I don't think I have `zh-CN` because I am using a new laptop after my old HDD has been corrupted :( but I might need to create a new tenant then test it again

**Comment by jtkech** [comment]
Okay cool then, nothing urgent but now I would like to understand what happens ;)

**Comment by hishamco** [comment]
Me too I will check tomorrow it's 3:17 AM here ;)

**Comment by jtkech** [comment]
> I don't think I have zh-CN because I am using a new laptop after my old HDD has been corrupted

Here I'm talking about the supported cultures coming from the Admin culture settings.

> but I might need to create a new tenant then test it again

Before, the 1st test that you could do in the exact same context is to put a breakpoint in the module startup just after we use our `AddSupportedCultures()`, to check if `zh-CN` was added in the supported cultures.

**Comment by hishamco** [comment]
> Before, the 1st test that you could do in the exact same context is to put a breakpoint in the module startup just after we use our AddSupportedCultures(), to check if zh-CN was added in the supported cultures.

Ya `zh-CN` in the document for unknown reason :) so, I understand now why the Culture Alias is not working because the culture is fallbacked to the default one, right?

**Comment by jtkech** [comment]
@hishamco 

Because the last time the document was persisted `zh-CN` was still available, to remove it from the supported culture I think you just need to edit again the culture settings and save them.

I could simulate it by adding manually `zh-CN` in the database document directly, in that case `zh-CN` is selected and mapped as for you, which is no more the case if I edit and save again the culture settings.

So as I was assuming, the aspnet middleware can only select a culture from the supported cultures (which includes the default one), so our next alias middleware can only map such a culture, which may happen for `zh-CN` if already in the database, but no longer the case when saving again the settings or on a new setup.

---

That's why for testing in #13131, if `zh-Hans-CN` is in the supported cultures, we also add dynamically `zh-CN` if not already there, so that it works for those that already have it or not after saving again the settings.

Then, if the aspnet middleware already set the culture to `zh-CN`, why mutating again the current culture by mapping it to `zh-Hans-CN`, we can discuss about what is better to do but it seems that using the `zh-CN` alias name still load a valid culture, the only problem being that it is no more listed in the admin UI.

That's why in #13131 we just add dynamically `zh-CN` and `zh-TW` and no longer use the alias middleware.

---

I'm open to any suggestion, but if we use the above solution (related to #13131), we would not need #13134 and #13143 to update the alias mapping that we would not use anymore.

We would also need to keep the `zh-CN` and `zh-TW` localization folders, even if we add the `zh-Hans-CN` and `zh-Hant-TW` folders. And we may need to re-consider #13076 to keep `zh-CN` and `zh-TW` in the setup cultures, even if we add the `zh-Hans-CN` and `zh-Hant-TW` cultures.

**Comment by hishamco** [comment]
> we can discuss about what is better to do but it seems that using the zh-CN alias name still load a valid culture, the only problem being that it is no more listed in the admin UI.

I'm agree to add them into admin UI, that's why I have the PR #13047

> We would also need to keep the zh-CN and zh-TW localization folders, even if we add the zh-Hans-CN and zh-Hant-TW folders.

I don't think they will remains after the last changes in the `OC.Translations` repo, but once the fallback is done they will be use `zh-Hans` and `zh-Hant` respectively
